src/appleseed.python/bindmasterrenderer.cpp

```diff
@@ -45,57 +45,100 @@ using namespace renderer;
+    // destroyed by python before the MasterRenderer is destroyed.
+    struct MasterRendererWrapper
+    {
+        MasterRendererWrapper(
+            bpy::object                 project,
+            const ParamArray&           params,
+            IRendererController*        renderer_controller,
+            ITileCallbackFactory*       tile_callback_factory = 0)
+          : m_project(project)
+        {
+            Project* proj = bpy::extract<Project*>(project);
+            m_renderer.reset(
+                new MasterRenderer(
+                    *proj,
+                    params,
+                    renderer_controller,
+                    tile_callback_factory));
+        }
+
+        MasterRendererWrapper(
+            bpy::object                 project,
+            const ParamArray&           params,
+            IRendererController*        renderer_controller,
+            ITileCallback*              tile_callback)
+            : m_project(project)
+        {
+            Project* proj = bpy::extract<Project*>(project);
+            m_renderer.reset(
+                new MasterRenderer(
+                    *proj,
+                    params,
+                    renderer_controller,
+                    tile_callback));
+        }
+
+        bpy::object                     m_project;
+        std::auto_ptr<MasterRenderer>   m_renderer;
+    };
+
+    std::auto_ptr<MasterRendererWrapper> create_master_renderer(
+        bpy::object             project,
         const bpy::dict&        params,
         IRendererController*    renderer_controller)
     {
         return
-            std::auto_ptr<MasterRenderer>(
-                new MasterRenderer(
+            std::auto_ptr<MasterRendererWrapper>(
+                new MasterRendererWrapper(
                     *project,
                     bpy_dict_to_param_array(params),
                     renderer_controller));
     }
 
-    std::auto_ptr<MasterRenderer> create_master_renderer_with_tile_callback(
-        Project*                project,
+    std::auto_ptr<MasterRendererWrapper> create_master_renderer_with_tile_callback(
+        bpy::object             project,
         const bpy::dict&        params,
         IRendererController*    renderer_controller,
         ITileCallback*          tile_callback)
     {
         return
-            std::auto_ptr<MasterRenderer>(
-                new MasterRenderer(
+            std::auto_ptr<MasterRendererWrapper>(
+                new MasterRendererWrapper(
                     *project,
                     bpy_dict_to_param_array(params),
                     renderer_controller,
                     tile_callback));
     }
 
-    bpy::dict master_renderer_get_parameters(const MasterRenderer* m)
+    bpy::dict master_renderer_get_parameters(const MasterRendererWrapper* m)
     {
-        return param_array_to_bpy_dict(m->get_parameters());
+        return param_array_to_bpy_dict(m->m_renderer->get_parameters());
```

Comment:
You could even overload `MasterRendererWrapper::operator->` to make it more transparent and minimize the number of changes...